pages/data-providers/data-provider.jsx

```diff
@@ -1,7 +1,133 @@
-import React from 'react'
+import React, { useState, useEffect } from 'react'
+import { useRouter } from 'next/router'
 
 import DataProviderPageTemplate from 'templates/data-provider'
+import useDebouncedEffect from 'hooks/use-debounced-effect'
 
-const DataProviderPage = () => <DataProviderPageTemplate />
+let timeout = null
+
+// TODO: Pages should be just a thin layer of NextJS.
+//       Find a better place for it.
+const useDiscoveryService = (defaultDataProviderUrl) => {
+  const [dataProviderUrl, setDataProviderUrl] = useState(defaultDataProviderUrl)
+  const [discoveredUrls, setDiscoveredUrls] = useState(null)
+  const [isLoading, setIsLoading] = useState(false)
+
+  const handleDataProviderUrlChange = (newUrl) => {
+    setDiscoveredUrls(null)
+    setDataProviderUrl(newUrl)
+  }
+
+  useDebouncedEffect(() => {
+    if (dataProviderUrl === '') return
+
+    setIsLoading(true)
+
+    // TODO: Call API Here
+    clearTimeout(timeout)
+    timeout = setTimeout(() => {
+      if (dataProviderUrl === 'success.url')
+        setDiscoveredUrls([{ url: dataProviderUrl, isNew: true }])
+      else if (dataProviderUrl === 'duplicate.url')
+        setDiscoveredUrls([{ url: dataProviderUrl }])
+      else if (dataProviderUrl === 'multiple-duplicates.url')
+        setDiscoveredUrls([{ url: dataProviderUrl }, { url: dataProviderUrl }])
+      else setDiscoveredUrls([])
+
+      setIsLoading(false)
+    }, 1000)
+  }, [dataProviderUrl])
+
+  useEffect(() => {
+    setDataProviderUrl(defaultDataProviderUrl)
+  }, [defaultDataProviderUrl])
+
+  return [
+    dataProviderUrl,
+    discoveredUrls,
+    isLoading,
+    handleDataProviderUrlChange,
+  ]
+}
+
+const getHelperMessage = (dataProviderUrl, discoveredUrls) => {
+  if (dataProviderUrl === '' || discoveredUrls === null) {
+    return {
+      message:
+        'It can be any URL: homepage address, any data resource address,' +
+        ' OAI-PMH endpoint or RIOXX endpoint',
+      variant: 'normal',
+    }
+  }
+
+  if (discoveredUrls.length === 0) {
+    return {
+      message:
+        'We could not find any OAI-PMH endpoint.' +
+        ' Please try it with another URL.',
+      variant: 'error',
+    }
+  }
```

Comment:
I would design it differently. As I said, expect from the API:
- 201 with the created entity (**Object**)
- 400 with the list of duplicates (**Array**)

I understand you wanted to DRY but it's confusing here. Maybe the store should be like:

```
{
  query: string, // user input
  request: Promise, // may be needed to cancel previous requests if the user inputs
  loading: boolean, // may be computed from the 'request'
  created: DataProvider,
  duplicated: Array(DataProvider), // matched, errored 🤔
  error: Error, // to handle the server error, like 500
}
```

It's just thinking, feel free to ignore it or put your hand here and edit to what you think is appropriate.

Reply:
I haven't really think about it. I wanted to see the response first. However proposed store looks good to me and it makes sense to structure it like that.